File: client/src/main/js/components/CourseStatisticsRefresh.js
import React from 'react';
import {Button} from 'reactstrap';
import ReactDOM from 'react-dom';
import {credentials} from '../scripts';
import {Api} from '../Api';
import {Notification} from './Notification';

export class CourseStatisticsRefresh extends React.Component {

    constructor(props) {
        super(props);

        this.synchronize = this.synchronize.bind(this);
    }

    synchronize() {
        Api.syncCourse(credentials(), this.props.course.name,
            () => ReactDOM.render(
                <Notification succeed
                              message={`Course ${this.props.course.name} synchronization
                                                has finished.`}/>,
                document.getElementById('notifications')
            ),
            response => ReactDOM.render(
                <Notification failed
                              message={`Course ${this.props.course.name} synchronization
                                                has failed due to: ${response}`}/>,
                document.getElementById('notifications')
            )
        )
    }

    render() {
        return (
            <Button color="info" outline className="icon-btn"
                    onClick={this.synchronize}
                    disabled={this.props.course.state.lifecycle !== 'RUNNING' ? 'disabled' : ''}>
                <i className="material-icons">refresh</i>
            </Button>
        );
    }
}

File: frontend/src/main/kotlin/org/flaxo/frontend/component/CourseStatisticsRefresh.kt
package org.flaxo.frontend.component

import kotlinx.coroutines.experimental.launch
import kotlinx.html.js.onClickFunction
import org.flaxo.frontend.Container
import org.flaxo.frontend.credentials
import org.flaxo.common.data.Course
import org.flaxo.common.data.CourseLifecycle
import org.flaxo.frontend.Notifications
import org.flaxo.frontend.client.FlaxoHttpException
import react.RBuilder
import react.dom.button
import react.dom.i

/**
 * Adds course statistics refresh button.
 */
fun RBuilder.courseStatisticsRefresh(course: Course) {
    button(classes = "btn btn-outline-info icon-btn") {
        attrs {
            onClickFunction = { launch { synchronizeCourseStatistics(course) } }
            disabled = course.state.lifecycle == CourseLifecycle.INIT
        }
        i(classes = "material-icons") { +"refresh" }
    }
}

private suspend fun synchronizeCourseStatistics(course: Course) {
    credentials?.also {
        try {
            Notifications.info("Course statistics refreshing was initiated.")
            Container.flaxoClient.syncCourse(it, course.name)
            Notifications.success("Course statistics synchronization has been finished.")
        } catch (e: FlaxoHttpException) {
            console.log(e)
            Notifications.error("Error occurred during ${course.name} course statistics synchronization.", e)
        }
    }
}
